### PR TITLE
Initial release of PingFederate

### PR DESCRIPTION
**Pull request by muskan-agarwal26**
## Proposed commit message

Create New integration package ping_federate.

- Added admin and audit data stream.
- Added data collection logic for the data streams.
- Added the ingest pipeline for the data streams.
- Mapped fields according to the ECS schema and added Fields metadata in the appropriate yml files.
- Added dashboards and visualizations.
- Added pipeline test for the data streams.
- Added system test cases for the data streams.

## Checklist

- [ ] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [ ] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).
- [ ] I have verified that any added dashboard complies with Kibana's [Dashboard good practices](https://docs.elastic.dev/ux-guidelines/data-viz/dashboard-good-practices) 

## How to test this PR locally

- Clone integrations repo.
- Install elastic package locally.
- Start elastic stack using elastic-package.
- Move to integrations/packages/ping_federate directory.
- Run the following command to run tests.

> elastic-package test
```
elastic-package test
Run asset tests for the package
2024/12/16 16:45:28  INFO License text found in "/root/github/integrations/LICENSE.txt" will be included in package
--- Test results for package: ping_federate - START ---
╭───────────────┬─────────────┬───────────┬────────────────────────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE       │ DATA STREAM │ TEST TYPE │ TEST NAME                                                              │ RESULT │ TIME ELAPSED │
├───────────────┼─────────────┼───────────┼────────────────────────────────────────────────────────────────────────┼────────┼──────────────┤
│ ping_federate │             │ asset     │ dashboard ping_federate-4e1a3b90-8da8-4d5d-89e7-5c49fb4a3541 is loaded │ PASS   │      1.538µs │
│ ping_federate │             │ asset     │ dashboard ping_federate-c975ed3a-3c11-4352-830e-84f1efaf9e92 is loaded │ PASS   │        256ns │
│ ping_federate │ admin       │ asset     │ index_template logs-ping_federate.admin is loaded                      │ PASS   │        150ns │
│ ping_federate │ admin       │ asset     │ ingest_pipeline logs-ping_federate.admin-0.1.0 is loaded               │ PASS   │        200ns │
│ ping_federate │ audit       │ asset     │ index_template logs-ping_federate.audit is loaded                      │ PASS   │        214ns │
│ ping_federate │ audit       │ asset     │ ingest_pipeline logs-ping_federate.audit-0.1.0 is loaded               │ PASS   │        214ns │
╰───────────────┴─────────────┴───────────┴────────────────────────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: ping_federate - END   ---
Done
Run pipeline tests for the package
--- Test results for package: ping_federate - START ---
╭───────────────┬─────────────┬───────────┬────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE       │ DATA STREAM │ TEST TYPE │ TEST NAME                                  │ RESULT │ TIME ELAPSED │
├───────────────┼─────────────┼───────────┼────────────────────────────────────────────┼────────┼──────────────┤
│ ping_federate │ admin       │ pipeline  │ (ingest pipeline warnings test-admin.log)  │ PASS   │  557.86265ms │
│ ping_federate │ admin       │ pipeline  │ test-admin.log                             │ PASS   │  214.58294ms │
│ ping_federate │ audit       │ pipeline  │ (ingest pipeline warnings test-audit.json) │ PASS   │ 494.078565ms │
│ ping_federate │ audit       │ pipeline  │ test-audit.json                            │ PASS   │ 210.997341ms │
╰───────────────┴─────────────┴───────────┴────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: ping_federate - END   ---
Done
Run policy tests for the package
--- Test results for package: ping_federate - START ---
No test results
--- Test results for package: ping_federate - END   ---
Done
Run static tests for the package
--- Test results for package: ping_federate - START ---
╭───────────────┬─────────────┬───────────┬──────────────────────────┬────────┬──────────────╮
│ PACKAGE       │ DATA STREAM │ TEST TYPE │ TEST NAME                │ RESULT │ TIME ELAPSED │
├───────────────┼─────────────┼───────────┼──────────────────────────┼────────┼──────────────┤
│ ping_federate │ admin       │ static    │ Verify sample_event.json │ PASS   │ 144.542411ms │
│ ping_federate │ audit       │ static    │ Verify sample_event.json │ PASS   │ 145.240233ms │
╰───────────────┴─────────────┴───────────┴──────────────────────────┴────────┴──────────────╯
--- Test results for package: ping_federate - END   ---
Done
Run system tests for the package
2024/12/16 16:45:37  INFO License text found in "/root/github/integrations/LICENSE.txt" will be included in package
2024/12/16 16:46:36  INFO Write container logs to file: /root/github/integrations/build/container-logs/ping_federate-logfile-admin-1734347796053043713.log
2024/12/16 16:47:30  INFO Write container logs to file: /root/github/integrations/build/container-logs/ping_federate-logfile-audit-1734347850709219074.log
2024/12/16 16:48:30  INFO Write container logs to file: /root/github/integrations/build/container-logs/ping_federate-tcp-audit-1734347910021317050.log
2024/12/16 16:49:27  INFO Write container logs to file: /root/github/integrations/build/container-logs/ping_federate-udp-audit-1734347967879909900.log
2024/12/16 16:50:24  INFO Write container logs to file: /root/github/integrations/build/container-logs/ping_federate-tls-audit-1734348024794769549.log
--- Test results for package: ping_federate - START ---
╭───────────────┬─────────────┬───────────┬───────────┬────────┬───────────────╮
│ PACKAGE       │ DATA STREAM │ TEST TYPE │ TEST NAME │ RESULT │  TIME ELAPSED │
├───────────────┼─────────────┼───────────┼───────────┼────────┼───────────────┤
│ ping_federate │ admin       │ system    │ logfile   │ PASS   │   38.8769948s │
│ ping_federate │ audit       │ system    │ logfile   │ PASS   │ 36.327029568s │
│ ping_federate │ audit       │ system    │ tcp       │ PASS   │ 40.646607107s │
│ ping_federate │ audit       │ system    │ tls       │ PASS   │ 39.845334757s │
│ ping_federate │ audit       │ system    │ udp       │ PASS   │ 39.845734973s │
╰───────────────┴─────────────┴───────────┴───────────┴────────┴───────────────╯
--- Test results for package: ping_federate - END   ---
Done
``` 

## Screenshots
pf-integration-page
pf-overview

### CODE REVIEW

File: packages/ping_federate/data_stream/admin/elasticsearch/ingest_pipeline/default.yml [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+            value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag {{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.on_failure_pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
+  - set:
+      field: event.kind
+      value: event
+      tag: set_event_kind
+  - append:
+      field: event.category
+      value: [authentication]
+      if: ctx.ping_federate?.admin?.event?.type != null
+      tag: append_event_category_authentication
```
Comment (kcreddy) [inline packages/ping_federate/data_stream/admin/elasticsearch/ingest_pipeline/default.yml:52]: Also, not all the events described in https://docs.pingidentity.com/pingfederate/latest/administrators_reference_guide/pf_admin_audit_loggin.html falls under `authentication`. 
From that list, only `Sign on attempt`, `Explicit user logout` and `Password change or reset` seem to be `authentication` related.

File: packages/ping_federate/data_stream/admin/elasticsearch/ingest_pipeline/default.yml [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+  - append:
+      field: event.category
+      value: [session]
```
Comment (kcreddy) [inline packages/ping_federate/data_stream/admin/elasticsearch/ingest_pipeline/default.yml:55]:
```suggestion
      value: session
```

Comment (kcreddy) [reply packages/ping_federate/data_stream/admin/elasticsearch/ingest_pipeline/default.yml:55]: Is there a change in Crest template? May I know why we are appending array to an array? 

Same for other processors in both datastreams.

File: packages/ping_federate/data_stream/admin/elasticsearch/ingest_pipeline/default.yml [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+      if: ctx.ping_federate?.admin?.event?.type != null && ctx.ping_federate?.admin?.event?.type.contains('SESSION')
+      tag: append_event_category_session
+  - append:
+      field: event.type
+      value: [info]
+      tag: append_event_type_info
+      if: >-
+        ctx.ping_federate?.admin?.event?.type != null && (
+          ctx.ping_federate?.admin?.event?.type == 'MODIFY' ||
```
Comment (kcreddy) [inline packages/ping_federate/data_stream/admin/elasticsearch/ingest_pipeline/default.yml:64]:
```suggestion
          ctx.ping_federate.admin.event.type == 'MODIFY' ||
```
Once checked for not-null, `?` can be ignored.

Same for other processors in both datastreams.

File: packages/ping_federate/data_stream/admin/elasticsearch/ingest_pipeline/default.yml [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        - yyyy-MM-dd H:mm:ss,SSS
+      if: ctx.ping_federate?.admin?.timestamp != null && ctx.ping_federate.admin.timestamp != ''
+      on_failure:
+        - remove:
+            field: ping_federate.admin.timestamp
+        - append:
+            field: error.message
+            value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag {{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.on_failure_pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
+# The timezone doesn't function as expected with custom date formats like yyyy-MM-dd HH:mm:ss,SSS,
+# Hence, first converted the date to ISO8601 format and then applied the event.timezone to the ISO8601 formatted date.
```
Comment (kcreddy) [inline packages/ping_federate/data_stream/admin/elasticsearch/ingest_pipeline/default.yml:105]: Is there a related public issue to this problem? You should link it.

File: packages/ping_federate/data_stream/admin/elasticsearch/ingest_pipeline/default.yml [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+      tag: set_pipeline_error_into_event_kind
+      value: pipeline_error
+      if: ctx.error?.message != null
+on_failure:
+  - append:
+      field: error.message
+      value: >-
+        Processor '{{{ _ingest.on_failure_processor_type }}}'
+        {{{#_ingest.on_failure_processor_tag}}}with tag '{{{ _ingest.on_failure_processor_tag }}}'
+        {{{/_ingest.on_failure_processor_tag}}}failed with message '{{{ _ingest.on_failure_message }}}'
```
Comment (kcreddy) [inline packages/ping_federate/data_stream/admin/elasticsearch/ingest_pipeline/default.yml:251]: Can you append `preserve_original_event` to `tags` inside this `on_failure` clause?
Example PR: https://github.com/elastic/integrations/pull/12046/files

@piyush-elastic , please update your templates (if any) as this change should be for all packages going forward.

File: packages/ping_federate/data_stream/admin/manifest.yml [hunk 6]
```diff
@@ -0,0 +1,57 @@
+title: Admin
```
Comment (kcreddy) [inline packages/ping_federate/data_stream/admin/manifest.yml:1]:
```suggestion
title: Admin logs
```

File: packages/ping_federate/data_stream/audit/elasticsearch/ingest_pipeline/default.yml [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+  - append:
+      field: related.hosts
+      value: '{{{ping_federate.audit.host.name}}}'
+      allow_duplicates: false
+      tag: append_related_hosts
+      if: ctx.ping_federate?.audit?.host?.name != null
+  - date:
+      field: cef.extensions.deviceReceiptTime
+      tag: date_ping_federate_audit_response_time
+      target_field: ping_federate.audit.response_time
```
Comment (kcreddy) [inline packages/ping_federate/data_stream/audit/elasticsearch/ingest_pipeline/default.yml:155]: May I know why `deviceReceiptTime` is being named as `response_time`? Do you mean `receipt_time`?

File: packages/ping_federate/manifest.yml [hunk 8]
```diff
@@ -0,0 +1,44 @@
+format_version: 3.2.1
+name: ping_federate
+title: PingFederate
+version: 0.1.0
+description: Collect logs from PingFederate with Elastic Agent.
+type: integration
+categories:
+  - security
```
Comment (kcreddy) [inline packages/ping_federate/manifest.yml:8]: Also add `authentication`

File: packages/ping_federate/data_stream/admin/elasticsearch/ingest_pipeline/default.yml (hunk 2 above)
Comment (muskan-agarwal26) [reply packages/ping_federate/data_stream/admin/elasticsearch/ingest_pipeline/default.yml:55]: This is by mistake, will resolve it everywhere.

File: packages/ping_federate/data_stream/audit/elasticsearch/ingest_pipeline/default.yml (hunk 7 above)
Comment (muskan-agarwal26) [reply packages/ping_federate/data_stream/audit/elasticsearch/ingest_pipeline/default.yml:155]: There is a field named : `responseTime` in audit, which is labelled as `deviceReceiptTime` in the CEF format, hence renamed it as `response_time`.
The fields for the audit log are mentioned in [this](https://github.com/pingidentity/pingidentity-server-profiles-playground/blob/master/getting-started-pingfederate-cluster/pingfederate/instance/server/default/conf/log4j2.xml) link

File: packages/ping_federate/_dev/build/docs/README.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+**[Audit](https://docs.pingidentity.com/pingfederate/latest/administrators_reference_guide/pf_security_audit_loggin.html)** - Provides a detailed record of authentication, authorization, and federation transactions.
+
+**Note**:
+
+1. In the Admin datastream, only logs from the admin.log file are supported via filestream in the pipe format. The log pattern is as follows:
+```
+<pattern>%d | %X{user} | %X{roles} | %X{ip} | %X{component} | %X{event} | %X{eventdetailid} | %m%n</pattern>
+```
+Sample Response:
```
Comment (kcreddy) [inline packages/ping_federate/_dev/build/docs/README.md:24]:
```suggestion
Sample Log:
```

File: packages/ping_federate/_dev/build/docs/README.md [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+```
+2024-11-28 5:58:55,832 | Administrator | UserAdmin,Admin,CryptoAdmin,ExpressionAdmin | 81.2.69.142 | A-rBnNPcJffxBiizBWDOWxq_Ek8cYxg3nxxxxyn6H4 | LICENSE | ROTATE | - Login was successful
+```
+
+2. Audit logs are supported through filestream, TCP, and UDP in the CEF format. The log pattern is as follows:
+```
+<pattern>%escape{CEF}{CEF:0|Ping Identity|PingFederate|%X{pfversion}|%X{event}|%X{event}|0|rt=%d{MMM dd yyyy HH:mm:ss.SSS} duid=%X{subject} src=%X{ip} msg=%X{status} cs1Label=Target Application URL cs1=%X{app} cs2Label=Connection ID cs2=%X{connectionid} cs3Label=Protocol cs3=%X{protocol} dvchost=%X{host} cs4Label=Role cs4=%X{role} externalId=%X{trackingid} cs5Label=SP Local User ID cs5=%X{localuserid} cs6Label=Attributes cs6=%X{attributes} %n}</pattern>
+```
+Sample Response:
```
Comment (kcreddy) [inline packages/ping_federate/_dev/build/docs/README.md:33]:
```suggestion
Sample Log:
```

File: packages/ping_federate/data_stream/admin/elasticsearch/ingest_pipeline/default.yml [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+      if: ctx.ping_federate?.admin?.timestamp != null && ctx.ping_federate.admin.timestamp != ''
+      on_failure:
+        - remove:
+            field: ping_federate.admin.timestamp
+        - append:
+            field: error.message
+            value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag {{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.on_failure_pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
+# The timezone doesn't function as expected with custom date formats like yyyy-MM-dd HH:mm:ss,SSS,
+# Hence, first converted the date to ISO8601 format and then applied the event.timezone to the ISO8601 formatted date.
+# An issue has been raised for this scenario here: https://discuss.elastic.co/t/timezone-offset-is-not-working-for-the-custom-date-formats/372010
```
Comment (kcreddy) [inline packages/ping_federate/data_stream/admin/elasticsearch/ingest_pipeline/default.yml:156]: Remove the link to [discuss](https://discuss.elastic.co) and create an issue in github repository and add that.

**Review by kcreddy — APPROVED** [review]

**Pull request merged**